### PR TITLE
Snapshots sources separately

### LINKED ISSUE #9980: Use input snapshots to skip empty work
There's no need to walk the inputs twice just to do this.

### PR DESCRIPTION
**Pull request by lptr**
Fixes #9980

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QF please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java [hunk 1]
```diff
@@ -21,70 +21,206 @@
...
+            work,
+            context.getAfterPreviousExecutionState()
+                .map(ExecutionState::getInputProperties)
+                .orElse(ImmutableSortedMap.of()),
+            context.getInputProperties(),
+            context.getInputFileProperties(),
+            (propertyName, type, identity) -> type.isSkipWhenEmpty());
+
+        // We don't currently have any non-file source properties
+        assert sourceInputs.getValueSnapshots().isEmpty();
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java:87]: Shall we throw an `IllegalStateException` or something like that here instead?

**Comment by wolfs** [comment]
Do we have a problem now that we treat the primary transform inputs as `skipWithEmpty`?

**Comment by lptr** [comment]
Superseded by https://github.com/gradle/gradle/pull/17837.

**Pull request closed without merging**